pip/commands/cache.py

```diff
@@ -0,0 +1,126 @@
+        if not args or args[0] not in self.actions:
+            raise CommandError(
+                "Please provide one of these subcommands: %s" %
+                ", ".join(self.actions)
+            )
+        method = getattr(self, "action_%s" % args[0])
+        return method(options, args[1:])
+
+    def get_cache_location(self, cache_root, cache_type):
+        location = cache_root
+        suffix = {"wheel": "wheels", "http": "http"}
+        if cache_type != "all":
+            location = os.path.join(location, suffix[cache_type])
+        return location
+
+    def action_info(self, options, args):
+        caches = ["http", "wheel"] if options.type == "all" else [options.type]
+        result = []
+        for cache_type in caches:
+            location = self.get_cache_location(options.cache_dir, cache_type)
+            stats = tree_statistics(location)
+            name = {"wheel": "Wheel cache", "http": "HTTP cache"}
+            result.append(textwrap.dedent(
+                """\
+                %s info:
+                   Location: %s
+                   Files: %s
+                   Size: %s""" %
+                (name[cache_type], location, stats["files"],
+                 format_size(stats["size"]))
+            ))
+        logger.info((os.linesep * 2).join(result))
+        return SUCCESS
+
+    def action_list(self, options, args):
+        if options.type != "wheel":
+            raise CommandError(
+                "pip cache list only operates on the wheel cache.")
+        cache_location = self.get_cache_location(options.cache_dir, "wheel")
+        wheels = [os.path.basename(f) for f in
+                  find_files(cache_location, "*.whl")]
+        wheels.sort()
+        logger.info(os.linesep.join(wheels))
+
+    def action_rm(self, options, args):
+        if options.type != "wheel":
+            raise CommandError(
+                "pip cache rm only operates on the wheel cache.")
+        if len(args) == 0:
+            raise CommandError(
+                "Must specify the filename of (a) wheel(s) to remove.")
+        cache_location = self.get_cache_location(options.cache_dir, "wheel")
+        cache_realpath = realpath(cache_location)
+        value = SUCCESS
+        for target in args:
+            matches = find_files(cache_location, target)
+            if not matches:
+                logger.warning("No match found for %s" % target)
+                value = ERROR
+                continue
+            for match in matches:
+                # protect against symlink attacks
+                if (os.path.commonprefix([cache_realpath, realpath(match)]) !=
+                        cache_realpath):
+                    logger.warning(
+                        "Did not remove %s because it is not a child of the "
+                        "wheel cache." % realpath(match))
+                    value = ERROR
+                    continue
+                try:
+                    os.unlink(match)
```

Comment:
On reflection, I think this symlink check narrows the window for a symlink attack race condition but doesn't eliminate it.

Let's assume our hapless user is getting rid of the wheels for a library called `passwd` by running `sudo pip cache rm passwd*`.

It's bad luck if an attacker manages to replace an intermediate directory of the cache with a symlink to /etc in the interval between `os.walk()` (in `find_files`) deciding to enter the directory and listing its contents.

This additional check permits a second race: after winning the first race, the attacker would have to make the intermediate directory a real file again before the check and then make it a symlink again before the os.unlink() call.

I think, rather than layering races, a better mitigation strategy is just to make sure the file looks like a wheel (i.e. matches the glob `*.whl`).